### ISSUE TITLE
Milestone v0.2 Parent Issue

### ISSUE THREAD

**Issue by nmushegian** [NexusDevelopment/dapple#42]
Milestone v0.2 
---

0.2 is a cleanup and review of the stuff we did to make dapple-js usable for real dev work. Work on 0.2 should be done with the vision for 0.3 in mind, also described here

* `dapple build`

collects sources form package dependency hierarchy and feeds it into solc. v0.2 does not need to deal with namespaces (re-throw solc's error if there are class name collisions), but does deal with aliasing packages so solidity imports work as expected.

Currently `build` emits 2 things:

*  a bundle containing all the output from `solc` that isn't totally redundant (filter `opcodes` and others) called `classes.json`
* a js module you can import to node directly or run through `browserify` in a UI package

`build` should export the types and objects described directly or indirectly in the dappfile. Right now that means it bundles the class definitions and interfaces. For 0.2 it should also ship a manually configured **set of named object with a classname (type)**, like a simple .json file in the `env` repo

* `dapple test`

run vm tests from build output against `testrpc` or any other "instant" eth rpc implementation.

### Looking ahead

0.3 will include `chain`, `env`, and `step` commands.

* `chain` is a process manager for clients which expose rpc/ipc interfaces used by dapple. When I do something in the context of `ethereum`, the `chain` module will tell me how to actually talk to `ethereum`

* `env` is for managing **environments**. An environment is a named set of object descriptions, which is an address and/or interface and/or definition (bytecode/source). Depending on what the module exports, people that import the package can refer to specific addresses in the deploy sequence, interact with deployed objects, or deploy new instances of defined classes.

* `step` takes a Step definition, an environment, and a chain connection, and applies the step to the blockchain, using the environment to resolve variables. An example is `deploy SomeContract($my_proxy_contact)`, where the environment would map `$my_proxy_contract` to an address and the chain would specify that you can talk to the chain at the default URL.

**Comment by mhhf** [comment]
I'd like to address some v0.3 features, but apart from the old documentation there no other resources to reason about the features of `chain`, `env` and `step`. What would help a lot, is when the cli (docopt and output?) along with the underlying data structures would be well defined. I just made a PR (#87) wich allows the use of [json-schemas](json-schema.org) for data specifications. This could make the breakdown of tasks easier and maybe attract newcomers like me and @drupalnomad.

E.g. for the **env** task the cli should look like this (taken and extended from [here](https://github.com/NexusDevelopment/dapple/blob/master/docs-old/usage-proposal.md))
```
    $ dapple env sync [<environment>]
    sync ok!

    $ dapple env new <name>
    > prompt about rpc-host, rpc-port, ipfs-host, ipfs-port
    New environment <name> added!

    $ dapple env ls
    <name1> @<host>:<port>
    ...

    $ dapple env rm <name>
    Environment <name> successfully removed
```
all manipulations affect the users [.dapplerc](https://github.com/mhhf/dapple/blob/feature/jsonSchema/specs/dapplerc.json)

* `new` should add a new [environment](https://github.com/mhhf/dapple/blob/feature/jsonSchema/specs/definitions.json#L98)
* `rm` removes a environment
* `ls` list all environments
* `sync` (???)

Also what do you think about working with a Kanban style task management? (with [waffle.io](https://waffle.io/NexusDevelopment/dapple)) or trello?

**Comment by ryepdx** [comment]
@nmushegian, I just realized that our thinking regarding `chain` and `env` might have diverged. Compare your descriptions of `env` and `chain` with what I wrote in [the usage proposal](https://github.com/NexusDevelopment/dapple/blob/master/docs-old/usage-proposal.md). In my view, chains and environments should probably be managed together, as they're intrinsically connected. But what are your thoughts on the matter?
